source/common/grpc/common.cc

```diff
@@ -74,19 +74,21 @@ absl::optional<Status::GrpcStatus> Common::getGrpcStatus(const Http::HeaderMap&
   //   1. trailers gRPC status, if it exists.
   //   2. headers gRPC status, if it exists.
   //   3. Inferred from info HTTP status, if it exists.
-  const std::array<absl::optional<Grpc::Status::GrpcStatus>, 3> optional_statuses = {{
-      {Grpc::Common::getGrpcStatus(trailers, allow_user_defined)},
-      {Grpc::Common::getGrpcStatus(headers, allow_user_defined)},
-      {info.responseCode() ? absl::optional<Grpc::Status::GrpcStatus>(
-                                 Grpc::Utility::httpToGrpcStatus(info.responseCode().value()))
-                           : absl::nullopt},
-  }};
-
-  for (const auto& optional_status : optional_statuses) {
-    if (optional_status.has_value()) {
-      return optional_status;
+  if (trailers) {
+    auto status = Grpc::Common::getGrpcStatus(*trailers, allow_user_defined);
```

Comment:
Minor perf optimizations to avoid unnecessary computation.

Reply:
I don't see any case outside of test passing nullptr to this? Did I miss anything?

Reply:
It's passed from context.cc which itself is used by Wasm context.cc. Wasm sets trailers to `nullptr` (a separate issue which we are debugging).

Reply:
IIRC right now we assume that headers are always passed to these functions, and static empty headers are passed in cases where there are none, which keeps the logic more simple. I would recommend you do the same from the WASM code unless there is a compelling reason to change this. If we do change it we should be consistent everywhere.

Reply:
Wasm captures local variables into object fields [here](https://github.com/envoyproxy/envoy-wasm/blob/master/source/extensions/common/wasm/context.h#L450), so it's natural to use `nullptr` as a default value (for performance reasons I assume to avoid de-allocating an optional). Generally speaking, I am not sure it's always right to replace missing headers with empty headers.

If there is a strong desire to use `HeaderMap&`, I can adapt the code to do that. WDYT?

Reply:
Originally I think the code was written this way to avoid having the conditional checks in a bunch of places. My point mainly is that I think we should be consistent here so I would take a look at all the callers.

Reply:
I reverted the changes outside of the context to keep the `nullptr` business confined to that unit. Using header map singleton consistent with other instances (e.g. `access_log_base.cc`).